### PR TITLE
Rework console dot's repository to something real 

### PR DESCRIPTION
**Pull request by jelly**
@martinpitt FYI

### CODE REVIEW

File: multiplexer.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            connection.request('POST', f'/v1.12/libpod/containers/{name}/start')
+            response = connection.getresponse()
+            print("podman start response:", response.status, response.reason)
+
+            # Add new entry to our sessions
+            SESSIONS[sessionid] = True
+
+            for sessionid in SESSIONS:
+                routes = f"""
+location /session/{sessionid} {{
```
Comment (martinpitt) [inline multiplexer.py:103]: location /api/cockpit-9090/session/{sessionid} {{

**Comment by martinpitt** [comment]
This is my current state: I can register two sessions, run two VMs, look up the random host port mapping in `sudo podman ps`, attach websocat/bridge in the VMs to the random host ports, and get two cockpit sessions on the session paths like https://localhost:8443/api/cockpit-9090/session/726d844e-ba10-4134-9a96-b846e42e90cb/ .

Problems:
 - Calling that URL without a trailing slash triggers an assertion in cockpit-ws
 - Creating the second session damaged the first one: I get "not a teapot" for all pages on the first session. This is a simple bug in the multiplexer, it does not write `location` stanzas for the old sessions, only the new one.

```diff
--- multiplexer.py
+++ multiplexer.py
@@ -76,6 +76,7 @@ class ProxyHTTPRequestHandler(BaseHTTPRequestHandler):
         print(self.path)
         if self.path == "/register":
             sessionid = str(uuid.uuid4())
+            # sessionid = "123"
             name = f'session-{sessionid}'
             connection = http.client.HTTPConnection('localhost')
             connection.sock = socket.socket(socket.AF_UNIX)
@@ -83,7 +84,9 @@ class ProxyHTTPRequestHandler(BaseHTTPRequestHandler):
             body = {
                     'image': 'quay.io/rhn_engineering_mpitt/ws',
                     'name': name,
-                    'command': ['sleep', 'infinity'],
+                    #'command': ['sleep', 'infinity'],
+                    'command': ['sh', '-exc', f"printf '[Webservice]\nUrlRoot=/api/cockpit-9090/session/{sessionid}/\\nWebsocketRoot=/wss/cockpit-9090/session/{sessionid}/\\nOrigins = https://localhost:8443\\n' > /etc/cockpit/cockpit.conf; exec /usr/libexec/cockpit-ws --for-tls-proxy --local-session=socat-session.sh"],
+                    'portmappings': [{ 'container_port': 8080}],
                     'remove': True,
                     'Networks': {'consoledot': {}},
             }
@@ -100,8 +103,52 @@ class ProxyHTTPRequestHandler(BaseHTTPRequestHandler):
 
             for sessionid in SESSIONS:
                 routes = f"""
-location /session/{sessionid} {{
-    proxy_pass http://session-{sessionid}:9090/api/cockpit-9090/{sessionid};
+location /api/cockpit-9090/session/{sessionid} {{
+    proxy_pass http://session-{sessionid}:9090/api/cockpit-9090/session/{sessionid};
+    # Don't pass on proxy auth
+    proxy_set_header Authorization "";
+
+    # Required to proxy the connection to Cockpit
+    proxy_set_header Host $host;
+    proxy_set_header X-Forwarded-Proto $scheme;
+
+    # Pass ETag header from Cockpit to clients.
+    # See: https://github.com/cockpit-project/cockpit/issues/5239
+    gzip off;
+}}
+location /wss/cockpit-9090/session/{sessionid} {{
+    proxy_pass http://session-{sessionid}:9090/wss/cockpit-9090/session/{sessionid};
+
+    # Required to proxy the connection to Cockpit
+    proxy_set_header Host $host;
+    proxy_set_header X-Forwarded-Proto $scheme;
+
+    # Required for web sockets to function
+    proxy_http_version 1.1;
+    proxy_buffering off;
+    proxy_set_header Upgrade $http_upgrade;
+    proxy_set_header Connection "upgrade";
+
+    # Pass ETag header from Cockpit to clients.
+    # See: https://github.com/cockpit-project/cockpit/issues/5239
+    gzip off;
+}}
+location /wss/cockpit-8080/session/{sessionid} {{
+    proxy_pass http://session-{sessionid}:8080/wss/cockpit-8080/session/{sessionid};
+
+    # Required to proxy the connection to Cockpit
+    proxy_set_header Host $host;
+    proxy_set_header X-Forwarded-Proto $scheme;
+
+    # Required for web sockets to function
+    proxy_http_version 1.1;
+    proxy_buffering off;
+    proxy_set_header Upgrade $http_upgrade;
+    proxy_set_header Connection "upgrade";
+
+    # Pass ETag header from Cockpit to clients.
+    # See: https://github.com/cockpit-project/cockpit/issues/5239
+    gzip off;
 }}
     """
             open('/etc/nginx/nginx.conf', 'w').write(NGINX_TEMPLATE.format(routes=routes))

```

File: Makefile [hunk 2]
```diff
@@ -1,21 +1,18 @@
...
+.PHONY: containers
+containers:
+	make -C 3scale container
+	make -C appservice container
 
-htpasswd:
-	htpasswd -bc htpasswd $(USER) $(PASSWORD)
+run: containers
+	podman network create $(NETWORK)
+	podman run --network $(NETWORK) --rm -p $(PORT):443 $(CONTAINER_NAME)
```
Comment (martinpitt) [inline Makefile:11]: what is CONTAINER_NAME here? also, shouldn't that run podman play kube etc?

File: Makefile [hunk 3]
```diff
@@ -1,21 +1,18 @@
...
 
-certs/service.pem:
-	mkdir -p certs && cd certs && sscg
+devel:
+	podman network create $(NETWORK)
+	podman run --name redis -d --network $(NETWORK) --rm redis:7.0.4
+	podman run --name multiplexer --network $(NETWORK) -d --rm -ti -p $(PORT):443 -v /run/podman/podman.sock:/run/podman/podman.sock -v $(PWD):/app $(CONTAINER_NAME) sleep infinity
```
Comment (martinpitt) [inline Makefile:16]: Is that still relevant?

File: appservice/multiplexer.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+import http.client
+import json
+import time
+
+import redis
+
+from multiprocessing import Process
+
+"""
+Only system containers!
```
Comment (martinpitt) [inline appservice/multiplexer.py:16]: Not any more! :stuck_out_tongue_winking_eye:

File: kube.yml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+---
```
Comment (martinpitt) [inline kube.yml:1]: Let's name it "webconsoledot-local.yaml" or so?

File: kube.yml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        - mountPath: /run/podman/podman.sock
+          name: podman-socket
+
+    - name: redis
+      image: docker.io/redis:7.0.4
+
+  volumes:
+    - name: podman-socket
+      hostPath:
+        path: /run/user/1000/podman/podman.sock
```
Comment (martinpitt) [inline kube.yml:28]: Hardcoding 1000 is "ugh", but let's fix that later.

File: multiplexer.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
+import os
```
Comment (martinpitt) [inline multiplexer.py:1]: We have this twice

**Review by martinpitt — CHANGES_REQUESTED** [review]

File: Makefile [hunk 8]
```diff
@@ -1,21 +1,18 @@
...
-	cat certs/service.pem certs/ca.crt > certs/service-chain.pem
+.PHONY: containers
+containers:
+	make -C 3scale container
+	make -C appservice container
 
-htpasswd:
-	htpasswd -bc htpasswd $(USER) $(PASSWORD)
+run: containers
+	podman network create $(NETWORK)
```
[the remark this answers is not in the document]
Comment (jelly) [reply Makefile:10]: I'd like to do that, but I can't remove a network if the pod is still around. So I'll settle with `network create || true` as podman play kube foo.yml --down does not remove anything

File: kube.yml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+      image: localhost/webconsoleapp:latest
+      ports:
+        - containerPort: 80
+          hostPort: 8080
+      volumeMounts:
+        - mountPath: /run/podman/podman.sock
+          name: podman-socket
+
+    - name: redis
+      image: docker.io/redis:7.0.4
```
[the remark this answers is not in the document]
Comment (jelly) [reply kube.yml:23]: Sure, but that mind be bound to change.

File: Makefile [hunk 10]
```diff
@@ -1,21 +1,16 @@
...
+	make -C 3scale container
+	make -C appservice container
 
 run:
-	podman run --rm -p $(PORT):443 $(CONTAINER_NAME)
+	podman network create $(NETWORK) || true
+	sed -e "s%XDG_RUNTIME_DIR%$${XDG_RUNTIME_DIR}%" webconsoledot-local.yaml | podman play kube --network $(NETWORK) -
+
+clean:
+	podman network rm -t0 -f $(NETWORK)
```
Comment (martinpitt) [inline Makefile:14]: --long --options please

**Review by martinpitt — CHANGES_REQUESTED** [review]

**Review by martinpitt — APPROVED** [review]
Dankjewel!

**Pull request merged**